pyop2/compilation.py

```diff
@@ -154,95 +157,121 @@ def compilation_comm(comm):
         can build object files and link in a single invocation, can be
-        overridden by exporting the environment variable ``LDSHARED``).
-    :arg cppargs: A list of arguments to the C compiler (optional, prepended to
-        any flags specified as the cflags configuration option)
-    :arg ldargs: A list of arguments to the linker (optional, prepended to any
-        flags specified as the ldflags configuration option).
+        overridden by exporting the environment variable ``PYOP2_LD``).
+    :arg extra_compiler_flags: A list of arguments to the C compiler (CFLAGS)
+        or the C++ compiler (CXXFLAGS)
+        ``PYOP2_CFLAGS`` and ``PYOP2_CXXFLAGS`` can also be used to extend these options
+        (optional, prepended to any flags specified as the cflags configuration option)
+    :arg extra_linker_flags: A list of arguments to the linker (LDFLAGS)
+        (optional, prepended to any flags specified as the ldflags configuration option).
     :arg cpp: Should we try and use the C++ compiler instead of the C
         compiler?.
     :kwarg comm: Optional communicator to compile the code on
         (defaults to COMM_WORLD).
     """
-    def __init__(self, cc, ld=None, cppargs=[], ldargs=[],
-                 cpp=False, comm=None):
-        ccenv = 'CXX' if cpp else 'CC'
+    _name = "unknown"
+
+    _cc = "mpicc"
+    _cxx = "mpicxx"
+    _ld = None
+
+    _cflags = []
+    _cxxflags = []
+    _ldflags = []
+
+    _optflags = []
+    _debugflags = []
+
+    def __init__(self, extra_compiler_flags=None, extra_linker_flags=None, cpp=False, comm=None, version=None):
+        self._extra_compiler_flags = extra_compiler_flags or []
+        self._extra_linker_flags = extra_linker_flags or []
+
+        self._cpp = cpp
+        self._debug = configuration["debug"]
+
         # Ensure that this is an internal communicator.
         comm = dup_comm(comm or COMM_WORLD)
         self.comm = compilation_comm(comm)
-        self._cc = os.environ.get(ccenv, cc)
-        self._ld = os.environ.get('LDSHARED', ld)
-        self._cppargs = cppargs + configuration['cflags'].split()
-        if configuration["use_safe_cflags"]:
-            self._cppargs += self.workaround_cflags
-        self._ldargs = ldargs + configuration['ldflags'].split()
+        self.version = version
+
+    def __repr__(self):
+        return f"<{self._name} compiler, version {self.ver or 'unknown'}>"
 
     @property
-    def compiler_version(self):
-        key = (id(self.comm), self._cc)
-        try:
-            return Compiler.compiler_versions[key]
-        except KeyError:
-            if self.comm.rank == 0:
-                ver = sniff_compiler_version(self._cc)
-            else:
-                ver = None
-            ver = self.comm.bcast(ver, root=0)
-            return Compiler.compiler_versions.setdefault(key, ver)
+    def cc(self):
+        return configuration["cc"] or self._cc
```

Comment:
Surely this is now redundant since we inspect `configuration["cc"]` earlier when deciding which subclass of `Compiler` to instantiate.